### PR TITLE
Made reclaim space task more tolerant in the face of artifact shared …

### LINKED ISSUE #3610: Reclaim space for repository fails with Cannot delete some instances of model 'Artifact' because they are referenced through protected foreign keys: 'ContentArtifact.artifact'."
**Version**
3.16, but probably all versions

**Describe the bug**

- Reclaim space for repository fails with the following error.

  ~~~
  Task paused with error: "("Cannot delete some instances of model 'Artifact' because they are referenced through protected foreign keys: 'ContentArtifact.artifact'.", {<ContentArtifact: pk=452959ad-c045-4e85-bf9f-6651ba37f57d>})"
  ~~~

**To Reproduce**
See BZ

**Additional context**
https://bugzilla.redhat.com/show_bug.cgi?id=2169322

### PR DESCRIPTION
**Pull request by ipanova**
…between two content units.

closes #3610

### CODE REVIEW

File: pulpcore/app/tasks/reclaim_space.py [hunk 1]
```diff
@@ -76,10 +82,16 @@ def reclaim_space(repo_pks, keeplist_rv_pks=None, force=False):
...
-        artifact.delete()
-        progress_bar.done += 1
-        counter += 1
+        try:
+            # we need to manually call delete() because it cleans up the file on the filesystem
+            artifact.delete()
+        except ProtectedError as e:
+            # Rarely artifact could be shared between to different content units.
+            # Just log and skip the artifact deletion in this case
+            log.info(e)
```
Comment (dralley) [inline pulpcore/app/tasks/reclaim_space.py:91]: Should this be a higher level such as WARN?

Comment (ipanova) [reply pulpcore/app/tasks/reclaim_space.py:91]: we could, but to be honest i would keep this at info as here https://github.com/pulp/pulpcore/blob/main/pulpcore/app/tasks/purge.py#L146 so we're consistent how we log same error

Comment (dralley) [reply pulpcore/app/tasks/reclaim_space.py:91]: ack

**Comment by dralley** [comment]
As an expedient workaround I'm perfectly OK with this.  I still want to figure out if we can address it on a more fundamental level though.

**Review by dralley — APPROVED** [review]

**Review by ggainey — APPROVED** [review]
Gotta love edge-cases...

File: pulpcore/app/tasks/reclaim_space.py [hunk 2]
```diff
@@ -76,10 +82,16 @@ def reclaim_space(repo_pks, keeplist_rv_pks=None, force=False):
...
+        else:
```
Comment (PaulSD) [inline pulpcore/app/tasks/reclaim_space.py:92]: Shouldn't this be "finally", not "else"?
The progress bar is tracking the position in `artifacts_to_delete`, so the `done` counter should be incremented when we move to the next artifact, even if this artifact was skipped / wasn't deleted.

Comment (ipanova) [reply pulpcore/app/tasks/reclaim_space.py:92]: @PaulSD  there was a little debate on how to treat such situation https://github.com/pulp/pulpcore/pull/4295#discussion_r1307250518
please read and let us know if you have any further feedback.

Comment (PaulSD) [reply pulpcore/app/tasks/reclaim_space.py:92]: I understand now.  This does make sense as-is.  Thanks!

**Pull request merged**